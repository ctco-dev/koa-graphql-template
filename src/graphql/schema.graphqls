# Root _Query_ definition
type Query {
  # Get _Joke_ by category
  jokeByCategory(category: JokeCategoryEnum): Joke
  # Get my _LifeStory_
  myLifeStory: String!
}

# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available joke's categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}
