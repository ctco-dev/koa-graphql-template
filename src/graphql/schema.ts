import { makeExecutableSchema } from 'graphql-tools';
import { nodeInterface, pageInfoType } from 'graphql-relay-tools';
import { importSchema } from 'graphql-import';

import resolvers from './resolvers';

import { swapiDef } from './schema/swapi';

const schemas = {
  nodeInterface,
  pageInfoType,
  ...swapiDef,
};

const typeDefs: string = importSchema('src/graphql/schema/schema.graphql', schemas);

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
