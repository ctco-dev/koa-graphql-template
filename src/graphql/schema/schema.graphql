# import * from "nodeInterface"
# import * from "pageInfoType"

# import Query.* from "jokes.graphql"
# import Query.* from "swapi"
