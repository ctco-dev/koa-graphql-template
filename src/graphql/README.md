GraphQL
=======

Currently the best approach seems to keep all schema definitions in a single file
 `schema.graphqls`. If it increases and becomes huge, probably it could be split into
 sensible parts and placed somewhere like `schemaDefs`. For merging schema pieces
 see [Modularizing the schema](http://dev.apollodata.com/tools/graphql-tools/generate-schema.html#modularizing).
 