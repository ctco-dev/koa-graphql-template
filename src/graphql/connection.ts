import { connectionDefinitions } from 'graphql-relay-tools';

/**
  * Constructs a GraphQL connection field config; it is assumed
  * that the object has a property named `prop`, and that property
  * contains a list of types.
  */
export function connectTypes(name: string, prop: string, type: string) {
  const { connectionType } = connectionDefinitions({
    name,
    nodeType: type,
    connectionFields: `
      totalCount: Int
      ${prop}: [${type}]
    `,
  });

  return connectionType;
}
