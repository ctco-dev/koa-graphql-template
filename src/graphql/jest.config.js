module.exports = Object.assign(
  {},
  require('../../tools/jest.default.config.js'),
  process.env.TEST_CI && require('../../tools/jest.ci.default.config.js'),
  {
    "setupFiles": [
      "./__tests__/__setup__/env-test.ts"
    ]
  }
);
