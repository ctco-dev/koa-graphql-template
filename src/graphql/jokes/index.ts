import * as fs from 'fs';
import * as path from 'path';

import resolvers from './resolvers';
import * as types from './types';

const schema: string = fs.readFileSync(path.join(__dirname, './jokes.graphqls'), 'utf8');

export {
  schema,
  resolvers,
  types,
};
