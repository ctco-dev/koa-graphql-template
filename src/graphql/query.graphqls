# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available joke's categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}

# Root _Query_ definition
type Query {
  # Get _Joke_ by category
  jokeByCategory(category: JokeCategoryEnum): Joke
}
