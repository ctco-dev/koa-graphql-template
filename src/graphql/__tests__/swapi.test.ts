import { graphql } from 'graphql';
import schema from '../schema';
import nock from 'nock';

const rootValue = {};
const context = {};

beforeAll(() => {

  // mock service endpoint

  const filmsResponse = {
    count: 2,
    next: null,
    previous: null,
    results: [
      {
        title: 'A New Hope',
        episode_id: 4,
        url: 'https://swapi.co/api/films/',
      },
      {
        title: 'Attack of the Clones',
        episode_id: 2,
        url: 'https://swapi.co/api/films/',
      },
    ],
  };

  nock(process.env.SWAPI_SERVICE_URL!)
    .get('/films/')
    .reply(200, filmsResponse);

  const filmResponse = {
    title: 'The Empire Strikes Back',
    episode_id: 5,
    url: 'https://swapi.co/api/films/2/',
  };

  nock(process.env.SWAPI_SERVICE_URL!)
    .get('/films/2/')
    .reply(200, filmResponse);
});

describe('query.swapi', () => {
  it('should match snapshot with pagination', async () => {
    const query = `
      query Q {
        allFilms {
          edges {
            node {
              title
              episodeID
            }
            cursor
          }
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          totalCount
          films {
            title
            episodeID
          }
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context);
    const { data, errors } = result;

    expect({ data, errors }).toMatchSnapshot();
  });

  it('should match snapshot without pagination', async () => {
    const query = `
      query Q {
        film(filmID: 2) {
          title
          episodeID
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context);
    const { data, errors } = result;

    expect({ data, errors }).toMatchSnapshot();
  });
});
