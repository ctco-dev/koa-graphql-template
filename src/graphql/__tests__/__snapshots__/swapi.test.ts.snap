// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query.swapi should match snapshot with pagination 1`] = `
Object {
  "data": Object {
    "allFilms": Object {
      "edges": Array [
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "node": Object {
            "episodeID": 4,
            "title": "A New Hope",
          },
        },
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
          "node": Object {
            "episodeID": 2,
            "title": "Attack of the Clones",
          },
        },
      ],
      "films": Array [
        Object {
          "episodeID": 4,
          "title": "A New Hope",
        },
        Object {
          "episodeID": 2,
          "title": "Attack of the Clones",
        },
      ],
      "pageInfo": Object {
        "endCursor": "YXJyYXljb25uZWN0aW9uOjE=",
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
      },
      "totalCount": 2,
    },
  },
  "errors": undefined,
}
`;

exports[`query.swapi should match snapshot without pagination 1`] = `
Object {
  "data": Object {
    "film": Object {
      "episodeID": 5,
      "title": "The Empire Strikes Back",
    },
  },
  "errors": undefined,
}
`;
